### PR TITLE
Add set rng functionality in preprocessor and model wrapper

### PR DESCRIPTION
**Pull request by dallasfoster**
This MR adds functionality to set the random seed from both the preprocessor class as well as the ModelWrapper class.

### CODE REVIEW

File: makani/models/preprocessor.py [hunk 1]
```diff
@@ -427,6 +427,13 @@ def get_internal_rng(self, gpu=True):
                 return self.input_noise.rng_cpu
         else:
             return None
+    
+    def set_rng(self, reset = True, seed=333):
```
Comment (bonevbs) [inline makani/models/preprocessor.py:431]: what is the logic behind the reset argument here? Shouldn't the userjust avoid calling set_rng if he does not want to overwrite it?

File: makani/models/model_package.py [hunk 2]
```diff
@@ -145,6 +145,14 @@ def out_channels(self):
     def timestep(self):
         return self.params.dt * self.params.dhours
 
+    def replace_state(self, replace_state=True):
```
Comment (azrael417) [inline makani/models/model_package.py:148]: Why do we pass a replace_state variable and do not use it later? We should either use it in the call or remove it. If we keep it, please rename that routine to update_state.

File: makani/models/model_package.py [hunk 3]
```diff
@@ -145,6 +145,14 @@ def out_channels(self):
...
+        self.model.preprocessor.update_internal_state(replace_state=True)
+        return
+    
+    def set_rng(self, reset = True, seed=333):
```
Comment (azrael417) [inline makani/models/model_package.py:152]: please remove the whitespace in the reset = True

File: makani/models/noise.py [hunk 4]
```diff
@@ -77,15 +77,18 @@ def __init__(
         self.mmax = self.isht.mmax
 
         # generator objects:
+        self.set_rng(seed=seed)
+
+        # store the noise state: initialize to None
+        self.register_buffer("state", torch.zeros((batch_size, self.num_time_steps, self.num_channels, self.lmax_local, self.mmax_local, 2), dtype=torch.float32), persistent=False)
```
Comment (azrael417) [inline makani/models/noise.py:83]: if se set persistent=False, the state will not be in the checkpoint? is this just for making it simpler to move it from device and back?

File: makani/models/noise.py [hunk 5]
```diff
@@ -99,23 +102,24 @@ def reset(self, batch_size=None):
...
-        # create single occurence
-        with torch.no_grad():
-            if batch_size is None:
-                batch_size = self.state.shape[0]
-            newstate = torch.empty((batch_size, self.num_time_steps, self.num_channels, self.lmax_local, self.mmax_local, 2), dtype=self.state.dtype, device=self.state.device)
-            if self.state.is_cuda:
-                newstate.normal_(mean=0.0, std=1.0, generator=self.rng_gpu)
-            else:
-                newstate.normal_(mean=0.0, std=1.0, generator=self.rng_cpu)
+         if replace_state:
```
Comment (azrael417) [inline makani/models/noise.py:105]: so the replace_state was not really used here, is that right? it seems correct to me now. did we never use this?

**Review by azrael417 — COMMENTED** [review]
lgtm, but had a few comments

File: makani/models/noise.py (hunk 4 above)
Comment (dallasfoster) [reply makani/models/noise.py:83]: Yes, the state will not be in the checkpoint but this was the original setting of this line. I have not changed it, only moved it.

Comment (azrael417) [reply makani/models/noise.py:83]: that sounds good, thanks

File: makani/models/noise.py (hunk 5 above)
Comment (azrael417) [reply makani/models/noise.py]: Ok, I see the issue. replace_state was introduced for stateful noise. So this is only relevant for diffusion noise sources. In this case, replace_state = True means that you want to overwrite what is in state. Now, for all other sources, update should always update the state. Can you undo this change? Or do you run into trouble when you do?

File: makani/models/preprocessor.py (hunk 1 above)
Comment (azrael417) [reply makani/models/preprocessor.py:431]: A call to update should always change the noise state, and for stateful noise we want to distinguish between generating a new noise source or doing an update.

Comment (azrael417) [reply makani/models/preprocessor.py:431]: I am indifferent about it. If the user knows what they are doing, I am happy with that.

File: makani/models/noise.py (hunk 5 above)
Comment (dallasfoster) [reply makani/models/noise.py]: Reverted.
